catalog/cataloger_merge.go

```diff
@@ -52,49 +54,46 @@ func (c *cataloger) Merge(ctx context.Context, repository, leftBranch, rightBran
 				Limit: -1,
 			},
 		}
-
-		scanner, err := NewDiffScanner(tx, params)
+		relation, err := getRefsRelationType(tx, params)
 		if err != nil {
 			return nil, err
 		}
-		if scanner.Relation == RelationTypeSame {
+		if relation == RelationTypeSame {
 			return nil, ErrSameBranchMergeNotSupported
 		}
 		nextCommitID, err := getNextCommitID(tx)
 		if err != nil {
 			return nil, err
 		}
-		// do the merge based on the relation
 		previousMaxCommitID, err := getLastCommitIDByBranchID(tx, rightID)
 		if err != nil {
 			return nil, err
 		}
-
-		differences := make(mergeBatchRecords, 0, MergeBatchSize)
+		mergeBatchChan := make(chan mergeBatchRecords, MergeBatchChanBuffer)
+		var workerExitFlag bool
+		go c.diffWorker(params, mergeBatchChan, &workerExitFlag)
```

Comment:
Consider - moving this code into a function/method that will make channels - one for differences and the second for an error. The function will make the channels, start the goroutine and defer close both channels when the worker is done.
We should pass a cancelation context based on the request context.
The worker should check the context while processing the diff scanners, on any error sending the error on an errors channel and quit.
The function will return the two channels to the caller and you can process the differences (batch or not batch in a transaction).
If the merge execution fails with an error, it would cancel the context given the diff.